File: .tmp_c4000/batch_04_go/3479_fruits_into_baskets_iii.go
// LeetCode 3479 - Fruits Into Baskets III
// https://leetcode.com/problems/fruits-into-baskets-iii/

func numOfUnplacedFruits(fruits []int, baskets []int) int {
	n := len(baskets)
	size := 1
	for size < n {
		size <<= 1
	}
	tree := make([]int, size*2)
	for i := 0; i < n; i++ {
		tree[size+i] = baskets[i]
	}
	for i := size - 1; i > 0; i-- {
		tree[i] = tree[i*2]
		if tree[i*2+1] > tree[i] {
			tree[i] = tree[i*2+1]
		}
	}
	var find func(int, int, int, int) int
	find = func(node, nl, nr, need int) int {
		if tree[node] < need {
			return -1
		}
		if nl == nr {
			return nl
		}
		mid := (nl + nr) / 2
		left := find(node*2, nl, mid, need)
		if left != -1 {
			return left
		}
		return find(node*2+1, mid+1, nr, need)
	}
	update := func(idx int) {
		p := size + idx
		tree[p] = -1
		for p >>= 1; p > 0; p >>= 1 {
			tree[p] = tree[p*2]
			if tree[p*2+1] > tree[p] {
				tree[p] = tree[p*2+1]
			}
		}
	}
	unplaced := 0
	for _, f := range fruits {
		idx := find(1, 0, size-1, f)
		if idx == -1 || idx >= n {
			unplaced++
		} else {
			update(idx)
		}
	}
	return unplaced
}

File: .tmp_py4013/batch_13_dump/3479_fruits_into_baskets_iii.js
// LeetCode 3479 - Fruits Into Baskets III
// https://leetcode.com/problems/fruits-into-baskets-iii/

var numOfUnplacedFruits = function(fruits, baskets) {
    const n = baskets.length;
    let size = 1;
    while (size < n) size <<= 1;
    const tree = new Array(size * 2).fill(0);
    for (let i = 0; i < n; i++) tree[size + i] = baskets[i];
    for (let i = size - 1; i > 0; i--) tree[i] = Math.max(tree[i * 2], tree[i * 2 + 1]);
    const find = (node, nl, nr, need) => {
        if (tree[node] < need) return -1;
        if (nl === nr) return nl;
        const mid = Math.floor((nl + nr) / 2);
        const left = find(node * 2, nl, mid, need);
        if (left !== -1) return left;
        return find(node * 2 + 1, mid + 1, nr, need);
    };
    const update = (idx) => {
        let p = size + idx;
        tree[p] = -1;
        for (p >>= 1; p > 0; p >>= 1) tree[p] = Math.max(tree[p * 2], tree[p * 2 + 1]);
    };
    let unplaced = 0;
    for (const f of fruits) {
        const idx = find(1, 0, size - 1, f);
        if (idx === -1 || idx >= n) unplaced++;
        else update(idx);
    }
    return unplaced;
};
